circuit-lib/src/elgamal-babyjub/elgamal.ts

```diff
@@ -0,0 +1,444 @@
+   *
+   * /// the circuit is loaded only once and it is available for use across multiple test cases.
+   * const circuit = await wasm_tester(path.resolve("./circuit/path"));
+   * const witness = await circuit.calculateWitness(inputsObject);
+   * await circuit.checkConstraints(witness);
+   * await circuit.loadSymbols();
+   *
+   * /// You can check signal names by printing "circuit.symbols".
+   * /// You will mostly need circuit inputs and outputs.
+   * const singalName = 'main.out'
+   * const signalValue = getSignalByName(circuit, witness, SignalName)
+   * ```
+   */
+  export const getSignalByName = (
+    circuit: any,
+    witness: any,
+    signalName: string,
+  ) => {
+    const signal = `main.${signalName}`;
+    return witness[circuit.symbols[signal].varIdx].toString();
+  };
+}
+
+export namespace DecodeUtils {
+  type LookupTable = { [key: string]: string };
+
+  const ProgressBar = require("cli-progress");
+  const bar = new ProgressBar.SingleBar({
+    format:
+      "Progress |" +
+      "{bar}" +
+      "| {percentage}% || {value}/{total} Chunks || Remaining time: {eta_formatted}",
+    barCompleteChar: "\u2588",
+    barIncompleteChar: "\u2591",
+    hideCursor: true,
+  });
+
+  /**
+   * Build a lookup table to break discrete log for 32-bit scalars for decoding
+   * @param precomputeSize the size of the lookup table to be used --> 2**precomputeSize
+   * @returns an object that contains 2**precomputeSize of keys and values
+   */
+  export function precompute(precomputeSize: number) {
+    const range = 32 - precomputeSize;
+    const upperBound = BigInt(2) ** BigInt(precomputeSize);
+
+    let lookupTable: LookupTable = {};
+    let key: string;
+
+    bar.start(Number(upperBound), 0);
+
+    for (let xhi = BigInt(0); xhi < upperBound; xhi++) {
+      key = babyJub.BASE.multiplyUnsafe(xhi * BigInt(2) ** BigInt(range))
+        .toAffine()
+        .x.toString();
+      lookupTable[key] = xhi.toString(16);
+      bar.update(Number(xhi) + 1);
+    }
+    bar.stop();
+
+    fs.writeFileSync(
+      `./src/elgamal-babyjub/lookupTableBBJub${precomputeSize}.json`,
+      JSON.stringify(lookupTable),
+    );
+  }
+
+  export function decode(
+    encoded: ExtPointType,
+    precomputeSize: number,
+    lookupTable: LookupTable,
+  ): bigint {
+    const range = 32 - precomputeSize;
+    const rangeBound = BigInt(2) ** BigInt(range);
+
+    for (let xlo = BigInt(0); xlo < rangeBound; xlo++) {
+      let loBase = babyJub.BASE.multiplyUnsafe(xlo);
+      let key = encoded.subtract(loBase).toAffine().x.toString();
+
+      if (lookupTable.hasOwnProperty(key)) {
+        return xlo + rangeBound * BigInt("0x" + lookupTable[key]);
+      }
+    }
+    throw new Error("Not Found!");
+  }
+
+  export function encode(plaintext: bigint): ExtPointType {
+    if (plaintext <= BigInt(2) ** BigInt(32)) {
+      return babyJub.BASE.multiplyUnsafe(plaintext);
+    } else throw new Error("The input should be 32-bit bigint");
```

Comment:
I prefer explicit errors because these are easier to read:
```
    if (plaintext > BigInt(2) ** BigInt(32)) {
    throw new Error("The input should be 32-bit bigint")
    }
    return babyJub.BASE.multiplyUnsafe(plaintext);
```